[Storage] Replace LoggingPolicy with @azure/logger and logPolicy

daviwil opened the pull request:
This change removes the storage libraries' `LoggingPolicy` implementation and replaces it with the new `logPolicy` from `core-http` which leverages `@azure/logging`.  The intent of this change is to connect the storage libraries with the standard logging pipeline so that its logs are consistent with those of the other Azure SDK libraries.

Obviously the implementation of the new `logPolicy` is simpler than what was there in `LoggingPolicy`, so we should definitely discuss what useful parts of it might be adopted into `core-http`.

Here's an example of what the log output looks like from `storage-blob` at the `info` level:

```
  azure:storage-blob:info Request: {
  azure:storage-blob:info   "streamResponseBody": false,
  azure:storage-blob:info   "url": "https://xxx.blob.core.windows.net/appendtestcontainer1/appendtestappendblob1?comp=REDACTED",
  azure:storage-blob:info   "method": "PUT",
  azure:storage-blob:info   "headers": {
  azure:storage-blob:info     "_headersMap": {
  azure:storage-blob:info       "content-type": "REDACTED",
  azure:storage-blob:info       "content-length": "REDACTED",
  azure:storage-blob:info       "x-ms-version": "REDACTED",
  azure:storage-blob:info       "user-agent": "REDACTED",
  azure:storage-blob:info       "x-ms-client-request-id": "3fcf4d38-da7e-47c1-a8d0-4f7e4202bf80"
  azure:storage-blob:info     }
  azure:storage-blob:info   },
  azure:storage-blob:info   "withCredentials": false,
  azure:storage-blob:info   "timeout": 0,
  azure:storage-blob:info   "keepAlive": true
  azure:storage-blob:info } +1ms
  azure:storage-blob:info Response status code: 201 +34ms
  azure:storage-blob:info Headers: {
  azure:storage-blob:info   "_headersMap": {
  azure:storage-blob:info     "content-length": "REDACTED",
  azure:storage-blob:info     "date": "REDACTED",
  azure:storage-blob:info     "etag": "REDACTED",
  azure:storage-blob:info     "last-modified": "REDACTED",
  azure:storage-blob:info     "server": "REDACTED",
  azure:storage-blob:info     "x-ms-blob-append-offset": "276",
  azure:storage-blob:info     "x-ms-blob-committed-block-count": "24",
  azure:storage-blob:info     "x-ms-client-request-id": "3fcf4d38-da7e-47c1-a8d0-4f7e4202bf80",
  azure:storage-blob:info     "x-ms-content-crc64": "REDACTED",
  azure:storage-blob:info     "x-ms-request-id": "d3809b03-001e-008e-4939-8577c8000000",
  azure:storage-blob:info     "x-ms-request-server-encrypted": "REDACTED",
  azure:storage-blob:info     "x-ms-version": "REDACTED"
  azure:storage-blob:info   }
  azure:storage-blob:info } +0ms
```

**Questions for Reviewers**
- Which headers and query params should we whitelist for each library?  I've added a couple for `storage-blob` since I was able to see them while running another test, but I'm not sure what would be needed for `storage-file` and `storage-queue`
- Does the [log level switching logic](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/storage/storage-blob/src/policies/LoggingPolicy.ts#L83-L90) actually help debugging in practice?  Is it worth considering this for inclusion in `core-http`'s `logPolicy`, perhaps configured by a setting?

Fixes #5628.

jeremymeng commented:
>     * Which headers and query params should we whitelist for each library?

I don't know the guideline for general headers but there's only a couple things we redact in storage.

jeremymeng commented on sdk/storage/storage-blob/src/Pipeline.ts:
```diff
@@ -200,7 +203,14 @@ export function newPipeline(
+      logger.info, {
+        allowedHeaderNames: [
+          "x-ms-blob-append-offset",
+          "x-ms-blob-committed-block-count"
```
I saw there's another PR with a set of default headers to enable.  Is this whitelist only for headers specific to Storage and to be determined by Storage team?

daviwil replied:
That PR was for the whitelist that works across all SDK libraries, the headers specified here are whitelisted in addition to the default list.

XiaoningLiu replied:
Prefer to move constant lists into constants.ts

---
In reply to: [336269928](https://github.com/Azure/azure-sdk-for-js/pull/5636#discussion_r336269928) [](ancestors = 336269928)

jeremymeng commented:
There are example code that uses console http logger in samples/readmes.  Okay to create a new issue to track updating them.

XiaoningLiu commented on sdk/storage/storage-queue/src/Pipeline.ts:
```diff
@@ -202,7 +206,10 @@ export function newPipeline(
-    new LoggingPolicyFactory()
+    logPolicy(
+      logger.info, {
+        allowedQueryParameters: ["api-version"]
```
>allowedQueryParameters: ["api-version"] [](start = 8, length = 39)

Do we have blacklist?

Following info should be skipped during logging:
- `sig` parameter in the url
- `Authentication` header in the header values

daviwil replied:
Anything not in the [default lists](https://github.com/Azure/azure-sdk-for-js/commit/d35927284640038bcf1f2d78135d95fbb46e965a) plus what is specified here is considered to be blacklisted, so the two you mentioned will show up as "REDACTED" in the logs.

XiaoningLiu replied:
I noticed there are other values are shown as REDACTED, for example x-ms-content-crc64, etag.

For storage, number of whitelist values is far larger than blacklisted values. Maintaining a short list of blacklist is more maintainable than a long whitelist list. Every time storage releases new API version, there will be new headers. And as protocol layer is auto generated, manually check new added headers every time sounds not perfect.

daviwil replied:
I believe the strategy we're going for across all SDK languages is to have an "allowed" list instead of a "deny" list just to avoid any unintended exposure of PII or secrets by default.  It does require more work on part of the service SDK teams to add new allowed items, but I think that's better than unintentionally causing sensitive data to be exposed when new service headers are added.

/cc @bterlson

bterlson replied:
Avoiding unintentional PII exposure is the high order bit here, and the allow list is the way to do it - it's too risky to make sure we're always on top of new headers/query params so we can sanitize them when required. See the [general guidelines](https://azure.github.io/azure-sdk/general_implementation.html#general-logging-requests) for more info.

daviwil commented:
@XiaoningLiu Checking back in on this -- are you OK with this approach in general?  I will take care of a couple of the comments you raised before.  We want to ship a trial-run of all the TS packages in the next couple of days and I'd like to have this merged to be included with that, if there are no strong objections.

Thanks a lot!

daviwil commented:
OK, this is ready for a final review pass.  A few changes since last time:

- I've added the complete allow-lists of header names and query parameters for provided by @tg-msft
- Moved `logger` definitions to their own file in each library and exported them from the public library interface
- Removed `LoggingPolicy` and `LoggingPolicyFactory` and updated API report files.  Some unrelated changes in these files are due to `api-extractor` sorting methods differently this run

daviwil commented:
Approved by adpship.
